---
title: Module 3
week: 3
---

* This will be a TOC
{:toc}

#Pascal's Triangle:

Authors

: Robert Sedgewick (book)

Take a look at the Wikipedia page on [Pascal's Triangle](http://en.wikipedia.org/wiki/Pascal%27s_triangle).  We will implement this using
a two-dimensional array, so it may be easier to imagine the entries left-justified rather than
in triangular form:

<PRE>
        column
        0  1  2  3  4
row
0       1
1       1  1
2       1  2  1
3       1  3  3  1
4       1  4  6  4  1
        .
        .
        .
</PRE>

Viewed as a two-dimensional array, the computation is specified
as follows, for each entry at row *r* and column *c*:

* If *c* is 0, then the entry is `1`.
* If *c*`==`*r*, then the entry is `1`.
* If *r*`<0` or *c*`<0` or *c*`>`*r*, then the entry is `0` and is not shown.
* Everywhere else, the entry is computed as the sum of the entries at:
	* *r*`-1`,*c*
	* *r*`-1`,*c*`-1`

##Procedure

1. Open the `PascalsTriangle` class, found in the
`arrays` package of the `extensions` folder.

2. Insert code to obtain from the user the value *N* which is the number of
rows you should compute of the triangle.

3. Instantiate the two-dimensional array needed to hold the results.

	>Java lets you do this as a *ragged* array, where each row can be of
	>a different length.  
	>
	>You are welcome to instantiate the array that way, but
	>it is simpler and equally acceptable to instantiate the array with the same 
	>number of columns in each row.  Java syntax for that is *much* simpler.
	>An example of this appears in `SelfAvoidingWalk`,
	>found in the
	>`book.ch1` package of the `book` source folder
	>in your repositories.

4. Compute the triangle as a two-dimensional array and print the results 
left-justified as shown above.

5. For extra fun, try to print the triangle centered as shown on
the [Wikipedia page](http://en.wikipedia.org/wiki/Pascal%27s_triangle).


#GC Content:
Authors

: [Ron K. Cytron](http://www.cs.wustl.edu/~cytron/)

Computer science plays a large role in modern biology.  BioJava is a framework for working with biological data.  It provides a framework for using sequences and performing common biology functions.  In this extension you will be reading in a DNA sequence and computing its 
GC-Content(http://en.wikipedia.org/wiki/GC-content).

###Notes
* You can find the Java files for this extension in the `extensions` source folder.

	>Locate the following files in the `biojavagc` package:
	> * `GCContent.java` is the file you complete.  When you run it,
	> analysis is performed on four genomes and the results are printed as output.
	>
	> * `GCContentTest.java` is a unit test for your work.  You will
	> receive credit for this extension only if this test passes.

* You can do this extension by following the instructions below.
However, you might want to take some time to familiarize yourself with the  [BioJava website](http://biojava.org/wiki/Main_Page).  
Of special importance is the [BioJava Cookbook](http://biojava.org/wiki/BioJava:CookBook).  The cookbook 
  has information and examples for performing common tasks.  You may also find the [BioJava API](http://www.biojava.org/docs/api/index.html) helpful.

* You will be working with some DNA sequences,
represented in the [FASTA format](http://en.wikipedia.org/wiki/FASTA_format),
which you can find in the `genomes` folder of your workspace:

	* `NC_002677.fasta`
	* `mysterya.fasta`
	* `mysteryb.fasta`
	* `mysteryc.fasta`

	> As you might guess from their names, there is an air of mystery surrounding
	> some of these genomes.
	> 
	> Consider a bacterium and a phage that might be hosted by that
	> bacterium.  It turns out that the DNA of the host and of the phage often need
	> to be similar in terms of their GC content for the bacterium to play host
	> to the phage:
	> 
	> Most bacteriophage and other bacteria are lower in GC content than Salmonella and its relatives, so invading DNA is an obvious target for H-NS. &quot;It's like a primitive immune system,&quot; says Fang. &quot;Reduce their expression, and the foreign genes can be tolerated.&quot;
	> 
	> From [http://www.ncbi.nlm.nih.gov/pmc/articles/PMC2064161/](http://www.ncbi.nlm.nih.gov/pmc/articles/PMC2064161/)

	In other words, if the GC content in the bacterium and phage is dissimilar, the bacterium may be
	immune to infection by the phage.

##Procedure
1. Complete the method `percentGC`  in `GCContent`.

	> Note: BioJava already has a function to get the GC-count (number of G and C nucleotides) of a sequence.  
	>
	> **Do not use this function.**
	>
	> Instead, devise a way to
	> iterate through the array of characters (representing nucleotides) yourself.  

2. Run the unit test `GCContentTest`  and make sure it passes.

	> Most students' code fails when given a string with no DNA in it whatsover
	> (an empty string or `""`.  Its GC content is 0%.

#Data Sorting and Analysis:
Authors
: Brennan Morell
: Jarett Gross

A big part of Computer Science revolves around finding ways
to sort data to be organized in a useful manner. There are many different algorithms
that computer scientists use to sort information. You can learn about a few popular ones below:


* [Insertion Sort](https://en.wikipedia.org/wiki/Selection_sort)
* [Merge Sort](https://en.wikipedia.org/wiki/Merge_sort)
* [Quick Sort](https://en.wikipedia.org/wiki/Quicksort)


> For the purposes of this assignment,
> we will implement our sorting using a naive process of repeatedly selecting the smallest value
> from a data set of type `int[]`, and swapping it to the front of the collection. In computer science,
> this is called a [Selection Sort](https://en.wikipedia.org/wiki/Insertion_sort).

##Procedure
1. Open the `Sorting` class, found in the `arrays` package
of the `extensions` folder.

2. Prompt the user for the `size` of the collection

	> If the user enters a negative number, continue to prompt them with a useful message until they enter a positive number.

3. Continue to prompt the user to input numbers one at a time, using an array to store the input data.

4. After processing all of the numbers entered, the data will be sorted using the following naive algorithm:

	> * Create an `int sortCount` to keep track of the number of times you've scanned the data
  	> 
	> * Create an outer loop to limit the number of scans to be equal to the size of the collection (`while(sortCount < size)`)
    > 
	> * Create an inner loop to iterate over the unsorted portion of the array, each time selecting the minimum value (keep track of this value and its index in the array)
	>
	> 	The unsorted portion will always be the sub-array with indexes `sortCount` to `size` inclusive.
	>
    > * Swap this minimum value with the value currently at the head of the unsorted portion of the array
    >
	> * At this point, `sortCount` should be incremented to reflect the new unsorted portion of the array.
	>
	> ![](http://www.piratelearner.com/static/media/images/admin/2015/10/13/selection.gif)

5. After the sorting the data, we will take advantage of its useful organization to compute the following statistics:

	Mean: The simple average of the data.

	Median: The middle value in the ordered dataset (Note: How does this computation vary for even and odd sized datasets?)

	Min: The smallest value in the ordered dataset.

	Max: The largest value in the ordered dataset.

	Range: The difference between the maximum and minimum values of the data set.

6. Finally, arrange for your output to display the sorted dataset and accompanying statistics in the following manner:

		1 2 3 4 5 
		
		Mean: 3.0
		
		Median: 3.0
		
		Min: 1
		
		Max: 5
		
		Range: 4

